File: common/src/main/java/net/polyv/android/player/common/ui/component/PLVMediaPlayerMoreActionLayoutLandscape.java
package net.polyv.android.player.common.ui.component;

import static net.polyv.android.player.sdk.foundation.lang.PreconditionsKt.requireNotNull;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;

import net.polyv.android.player.business.scene.common.model.vo.PLVMediaOutputMode;
import net.polyv.android.player.common.R;
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider;
import net.polyv.android.player.common.modules.media.viewmodel.PLVMPMediaViewModel;
import net.polyv.android.player.common.modules.media.viewmodel.viewstate.PLVMPMediaInfoViewState;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.PLVMPMediaControllerViewModel;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.viewstate.PLVMPMediaControllerFloatAction;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.viewstate.PLVMPMediaControllerViewState;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberState;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberStateCompareResult;

import java.util.List;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

/**
 * @author Hoshiiro
 */
public class PLVMediaPlayerMoreActionLayoutLandscape extends FrameLayout implements View.OnClickListener {

    private ImageView moreActionCloseIv;
    private ConstraintLayout moreActionContainer;
    private PLVMediaPlayerMoreLayoutAudioModeActionView moreAudioModeActionView;

    protected List<PLVMediaOutputMode> currentSupportMediaOutputModes = null;
    protected boolean isVisible = false;

    public PLVMediaPlayerMoreActionLayoutLandscape(@NonNull Context context) {
        super(context);
    }

    public PLVMediaPlayerMoreActionLayoutLandscape(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public PLVMediaPlayerMoreActionLayoutLandscape(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    {
        LayoutInflater.from(getContext()).inflate(R.layout.plv_media_player_ui_component_more_action_layout_landscape, this);
        moreActionCloseIv = findViewById(R.id.plv_media_player_more_action_close_iv);
        moreActionContainer = findViewById(R.id.plv_media_player_more_action_container);
        moreAudioModeActionView = findViewById(R.id.plv_media_player_more_audio_mode_action_view);

        setOnClickListener(this);
        moreActionCloseIv.setOnClickListener(this);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (isInEditMode()) return;

        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaViewModel.class)
                .getMediaInfoViewState()
                .observeUntilViewDetached(this, new Function1<PLVMPMediaInfoViewState, Unit>() {
                    @Override
                    public Unit invoke(PLVMPMediaInfoViewState viewState) {
                        currentSupportMediaOutputModes = viewState.getSupportOutputModes();
                        onViewStateChanged();
                        return null;
                    }
                });

        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaControllerViewModel.class)
                .getMediaControllerViewState()
                .observeUntilViewDetached(this, new Function1<PLVMPMediaControllerViewState, Unit>() {
                    @Override
                    public Unit invoke(PLVMPMediaControllerViewState viewState) {
                        isVisible = viewState.getLastFloatActionLayout() == PLVMPMediaControllerFloatAction.MORE;
                        onViewStateChanged();
                        return null;
                    }
                });
    }

    protected void onViewStateChanged() {
        PLVRememberState.rememberStateOf(this, "onSupportMediaOutputModesChanged")
                .compareLastAndSet(currentSupportMediaOutputModes)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult result) {
                        onSupportMediaOutputModesChanged();
                        return null;
                    }
                });

        PLVRememberState.rememberStateOf(this, "onChangeVisibility")
                .compareLastAndSet(isVisible)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult result) {
                        onChangeVisibility();
                        return null;
                    }
                });
    }

    protected void onSupportMediaOutputModesChanged() {
        if (currentSupportMediaOutputModes == null) {
            return;
        }
        boolean showAudioModeAction = currentSupportMediaOutputModes.contains(PLVMediaOutputMode.AUDIO_ONLY);
        moreAudioModeActionView.setVisibility(showAudioModeAction ? View.VISIBLE : View.GONE);
    }

    protected void onChangeVisibility() {
        setVisibility(isVisible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == moreActionCloseIv.getId()) {
            closeLayout();
        } else if (id == this.getId()) {
            closeLayout();
        }
    }

    private void closeLayout() {
        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaControllerViewModel.class)
                .popFloatActionLayout();
    }

}

File: common/src/main/java/net/polyv/android/player/common/ui/component/more/PLVMediaPlayerMoreActionLayoutLandscape.kt
package net.polyv.android.player.common.ui.component.more

import android.content.Context
import androidx.constraintlayout.widget.Guideline
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import net.polyv.android.common.libs.lang.state.watchStates
import net.polyv.android.player.business.scene.common.model.vo.PLVMediaOutputMode
import net.polyv.android.player.common.R
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider
import net.polyv.android.player.common.modules.media.viewmodel.PLVMPMediaViewModel
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.PLVMPMediaControllerViewModel
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.viewstate.PLVMPMediaControllerFloatAction

/**
 * @author Hoshiiro
 */
class PLVMediaPlayerMoreActionLayoutLandscape @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr), View.OnClickListener {

    private val moreActionStartGuideLine by lazy { findViewById<Guideline>(R.id.plv_media_player_more_action_start_guide_line) }
    private val moreAudioModeActionView by lazy { findViewById<PLVMediaPlayerMoreLayoutAudioModeActionView>(R.id.plv_media_player_more_audio_mode_action_view) }
    private val moreFloatWindowActionView by lazy { findViewById<PLVMediaPlayerMoreLayoutFloatWindowActionView>(R.id.plv_media_player_more_float_window_action_view) }
    private val moreSubtitleActionView by lazy { findViewById<PLVMediaPlayerMoreLayoutSubtitleActionViewLand>(R.id.plv_media_player_more_subtitle_action_view) }
    private val moreDownloadActionView by lazy { findViewById<PLVMediaPlayerMoreLayoutDownloadActionView>(R.id.plv_media_player_more_download_action_view) }
    private val moreKnowledgeActionView by lazy { findViewById<PLVMediaPlayerMoreLayoutKnowledgeActionView>(R.id.plv_media_player_more_knowledge_action_view) }
    private val moreActionCloseIv by lazy { findViewById<ImageView>(R.id.plv_media_player_more_action_close_iv) }

    init {
        LayoutInflater.from(context).inflate(R.layout.plv_media_player_ui_component_more_action_layout_landscape, this)
        setOnClickListener(this)
        moreActionCloseIv.setOnClickListener(this)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (isInEditMode) return

        val dependScope = PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!

        watchStates {
            val infoState = dependScope.get<PLVMPMediaViewModel>().mediaInfoViewState.value ?: return@watchStates
            val currentSupportMediaOutputModes = infoState.supportOutputModes
            val showAudioModeAction = currentSupportMediaOutputModes.contains(PLVMediaOutputMode.AUDIO_ONLY)
            moreAudioModeActionView.visibility = if (showAudioModeAction) VISIBLE else GONE
        }.disposeOnDetached(this)

        watchStates {
            val controllerState = dependScope.get<PLVMPMediaControllerViewModel>().mediaControllerViewState.value
            val isVisible = controllerState?.lastFloatActionLayout == PLVMPMediaControllerFloatAction.MORE
            visibility = if (isVisible) VISIBLE else GONE
        }.disposeOnDetached(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            moreActionCloseIv.id -> closeLayout()
            this.id -> closeLayout()
        }
    }

    private fun closeLayout() {
        PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
            .get<PLVMPMediaControllerViewModel>()
            .popFloatActionLayout()
    }
}
